spec/index.bs

```diff
@@ -1828,6 +1828,12 @@ A <dfn dfn>supported limits</dfn> object has a value for every limit defined by
     <tr class=row-continuation><td colspan=4>
         The maximum value for the arguments of
         {{GPUComputePassEncoder/dispatchWorkgroups(workgroupCountX, workgroupCountY, workgroupCountZ)}}.
+
+    <tr><td><dfn>maxMultiDrawIndirectCount</dfn>
```

Comment:
A few thoughts here:
- Shouldn't this also apply to the value in the `drawCountBuffer`?
- Does this mean that after enabling the feature you still need to request a higher limit or else will be stuck to 1? It's workable but seems a bit weird.
- Seems a bit weird that this could be 1 when the feature is not enabled. In fact it seems weird the limit exists at all when the feature isn't enabled...I don't think there precedent for feature related limits though.
- @kvark had suggested in a [previous comment](https://github.com/gpuweb/gpuweb/pull/2315#discussion_r801040476) to " lift the limit entirely...I.e. it becomes limited in the same way as maximum index count is limited." It's not clear to me how "maximum index count is limited", other than it's tied to the max buffer size. Maybe the feature should only be available when the implementation max count is >= `maxStorageBufferBindingSize` and then this limit goes away? Not sure how reasonable that is.

Reply:
> Shouldn't this also apply to the value in the drawCountBuffer?
Realized this is already covered transitively since drawCountBuffer is already limited by maxDrawCount. Nm.

Reply:
@Sirtsu55 is working on this feature in Dawn and we got to talking about the limit. I am not sure we need one - if it's guaranteed to be large on all systems then we could just hardcode a limit, or derive a limit from `maxBufferSize`.

- Vulkan: has a limit but it's [always either 1 or huge](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxDrawIndirectCount&platform=all), so when the feature is available we could just have a hardcoded limit as high as `2**30`.
- D3D12: not sure yet, currently investingating.
- Metal: There is already a maxBufferSize which imposes some limit. On weaker devices like old iPhones (the ones where maxBufferSize is 256MiB) there's a good chance this is the limiting factor. However the limiting factor wouldn't be the size on the indirect buffer (256MiB/16B=16.6M draws or 256MiB/20B=13.4M indexed draws) but the size of the indirect command buffer (ICB) that we generate internally in our draw call validation/rerecording shader. There is a [maxCommandCount](https://developer.apple.com/documentation/metal/mtldevice/2967420-newindirectcommandbufferwithdesc?language=objc) when you allocate an ICB, but we haven't found any documented limit on how large this can be. We don't know what the size of a draw command is so we can't seem to derive it from [maxBufferLength](https://developer.apple.com/documentation/metal/mtldevice/2966563-maxbufferlength?language=objc) - maybe in practice it's something like `maxBufferLength / 32 bytes` (256MiB/32B = 8.3M). We will probably try probing the Metal validation layer, but @mwyrzykowski @litherum would you be able to tell us how to determine this limit - maybe what the validation layer checks and/or whether it would be large enough on all target systems (#1069)?

Reply:
@kainino0x  The metal validation layer does not appear to validate maxCommandCount specifically, though you may run into validation errors due to the maxBufferLength. 

It appears we need the following to hold `device.maxBufferLength <= 40 + commandCount * ((2 + 2 * desc.maxVertexBufferBindCount + desc.maxFragmentBufferBindCount) * 8)`, where `desc` is the ICB descriptor.

I will need to check if that is strictly for validation or applies in general, we could start there as a limit.

Reply:
@mwyrzykowski thanks! Since multidraw will always inherit vertex/fragment buffers, those should be 0, simplifying to `40 + maxCommandCount * 2 * 8`. (Also, did you mean the reverse, `... <= maxBufferLength`?)

Just want to verify that 16 bytes per command is correct. That is too small for an indexed indirect call (20 bytes) so I'm guessing it is 2 pointers, one which points to the indexed indirect parameters allocated in some other buffer. If so, does the size of that buffer impose a limit? Or will it split into multiple buffers if it doesn't fit?

I'm also concerned that there will be three huge allocations, and we'll run out of memory anyway: the user's draw params, the ICB full of pointers, and Metal's internal allocation(s) that hold draw params.

Reply:
Oh yes it is reversed. Is it true that multi draw always inherits vertex/fragment buffers?

Wouldn't it be possible to use it from GPURenderBundleEncoder?

Reply:
Assuming the only hard limit on Metal is `maxCommandCount * 16 < maxBufferSize` we could maybe pick a hardcoded limit of ~16M (`floor(256MiB - 40B) / 16B) = 16777213`).

On D3D12, Dawn currently allocates a buffer with 28 bytes per draw (20B for the draw + 8B for the first vertex/index needed to provide correct vertex_index and instance_index builtins) - but there are no hard buffer size limits on internal buffers, so that should be OK, modulo running out of memory (the user has up to `maxDrawCount*20` bytes and we allocate up to `maxDrawCount*28` more, the latter of which is ~448MiB).

We could also pick a more comfortable smaller hardcoded limit and make applications split their multidraw calls if they really want huge numbers, but splitting is nontrivial.

Reply:
> Is it true that multi draw always inherits vertex/fragment buffers?

Yes. The ICB we are building is just for the output of a compute shader which validates and translates the multidraw parameters into an ICB full of draw calls and nothing else. We create a new ICB for each multidraw call (or possibly several multidraw calls batched together, like we do for regular draw calls, if we find we need to optimize that).

> Wouldn't it be possible to use it from GPURenderBundleEncoder?

It should be possible, yes. Is that a problem because the render bundle can itself be an ICB?

Reply:
> It should be possible, yes. Is that a problem because the render bundle can itself be an ICB?

Yes, and the ICB render command can not call executeBundles. Or would the use of multi-draw indirect in a GPURenderBundle force (partial) fallback to record+replay of commands?

Perhaps there is a way for ICB to support multi-draw indirect just as ICBs can support drawIndirect and drawIndexedIndirect via some emulation, as those commands are also missing from MTLRe, I just haven't thought of it off hand.

Reply:
This is a great question and I don't think we have an answer for it. Maybe it would be possible to expand multidraw calls in bundles out, into allocation space for many single-draw calls, which we can write to later? But that would leave (maxDrawCount - actual draw count) of unused draw calls in the bundle which is probably very inefficient.

Maybe we have to say multidraw is not supported in bundles to start.

Reply:
@kainino0x What should we do about the `maxMultiDrawIndirectCount` limit?

Reply:
Just to put it out there:

From my investigation of D3D12 in WebGPU the limiting factor is the buffer size. I tried counts up to 2^26 in a native D3D12 app, it does not work. Vulkan has a property and Metal is untested.

Reply:
I don't know, I'd propose giving a hardcoded value of 10M or something to punt again on adding limits that are part of features (#4613). Unless we think that will be too small.

Of course we need to test that, and we also need to check that setting maxDrawCount=10M and having a small draw count in the drawCountBuffer can perform passably well. (It should, though in Dawn we need to fix <https://crbug.com/367394543>.)